oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.auth.oauth2;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.auth.oauth2.AccessTokenCapsule;
+import com.linecorp.armeria.common.auth.oauth2.RefreshAccessTokenRequest;
+import com.linecorp.armeria.common.auth.oauth2.TokenRequestException;
+
+/**
+ * Base implementation of OAuth 2.0 Access Token Grant flow to obtain Access Token.
+ * Implements Access Token loading, storing and refreshing.
+ */
+abstract class AbstractOAuth2AuthorizationGrant implements OAuth2AuthorizationGrant {
+
+    /**
+     * A period when the token should be refreshed proactively prior to its expiry.
+     */
+    private static final Duration DEFAULT_REFRESH_BEFORE = Duration.ofMinutes(1L); // 1 minute
+
+    private final AtomicReference<AccessTokenCapsule> tokenRef;
+
+    private final RefreshAccessTokenRequest refreshRequest;
+
+    private final Duration refreshBefore;
+
+    @Nullable
+    private final Supplier<AccessTokenCapsule> tokenSupplier;
+    @Nullable
+    private final Consumer<AccessTokenCapsule> tokenConsumer;
+
+    AbstractOAuth2AuthorizationGrant(RefreshAccessTokenRequest refreshRequest, @Nullable Duration refreshBefore,
```

Comment:
How about making `refreshBefore` non-null here and moving the default to the builder?

Reply:
fixed